cmd/tsbs_generate_queries/databases/victoriametrics/devops.go

```diff
@@ -0,0 +1,161 @@
+}
+
+func (d *Devops) GroupByOrderByLimit(qi query.Query) {
+	panic("GroupByOrderByLimit not supported in PromQL")
+}
+
+func (d *Devops) LastPointPerHost(qq query.Query) {
+	panic("LastPointPerHost not supported in PromQL")
+}
+
+// GroupByTime selects the MAX for numMetrics metrics under 'cpu'
+// per minute for nhosts hosts,
+// e.g. in pseudo-PromQL:
+// max(
+// 	max_over_time(
+// 		{__name__=~"metric1|metric2...|metricN",hostname=~"hostname1|hostname2...|hostnameN"}[1m]
+// 	)
+// ) by (__name__)
+func (d *Devops) GroupByTime(qq query.Query, nHosts, numMetrics int, timeRange time.Duration) {
+	metrics := mustGetCPUMetricsSlice(numMetrics)
+	hosts := d.mustGetRandomHosts(nHosts)
+	selectClause := getSelectClause(metrics, hosts)
+	qi := &queryInfo{
+		query:    fmt.Sprintf("max(max_over_time(%s[1m])) by (__name__)", selectClause),
+		label:    fmt.Sprintf("VictoriaMetrics %d cpu metric(s), random %4d hosts, random %s by 1m", numMetrics, nHosts, timeRange),
+		interval: d.Interval.MustRandWindow(timeRange),
+		step:     "60",
+	}
+	d.fillInQuery(qq, qi)
+}
+
+// GroupByTimeAndPrimaryTag selects the AVG of numMetrics metrics under 'cpu' per device per hour for a day,
+// e.g. in pseudo-PromQL:
+//
+// avg(
+// 	avg_over_time(
+// 		{__name__=~"metric1|metric2...|metricN"}[1h]
+// 	)
+// ) by (__name__, hostname)
+//
+// Resultsets:
+// double-groupby-1
+// double-groupby-5
+// double-groupby-all
+func (d *Devops) GroupByTimeAndPrimaryTag(qq query.Query, numMetrics int) {
+	metrics := mustGetCPUMetricsSlice(numMetrics)
+	selectClause := getSelectClause(metrics, nil)
+	qi := &queryInfo{
+		query:    fmt.Sprintf("avg(avg_over_time(%s[1h])) by (__name__, hostname)", selectClause),
+		label:    devops.GetDoubleGroupByLabel("VictoriaMetrics", numMetrics),
+		interval: d.Interval.MustRandWindow(devops.DoubleGroupByDuration),
+		step:     "3600",
+	}
+	d.fillInQuery(qq, qi)
+}
+
+// MaxAllCPU selects the MAX of all metrics under 'cpu' per hour for nhosts hosts,
+// e.g. in pseudo-PromQL:
+//
+// max(
+// 	max_over_time(
+// 		{hostname=~"hostname1|hostname2...|hostnameN"}[1h]
+// 	)
+// ) by (__name__)
+func (d *Devops) MaxAllCPU(qq query.Query, nHosts int) {
+	hosts := d.mustGetRandomHosts(nHosts)
+	selectClause := getSelectClause(devops.GetAllCPUMetrics(), hosts)
+	qi := &queryInfo{
+		query:    fmt.Sprintf("max(max_over_time(%s[1h])) by (__name__)", selectClause),
+		label:    devops.GetMaxAllLabel("VictoriaMetrics", nHosts),
+		interval: d.Interval.MustRandWindow(devops.MaxAllDuration),
+		step:     "3600",
+	}
+	d.fillInQuery(qq, qi)
+}
+
+// HighCPUForHosts populates a query that gets CPU metrics when the CPU has high
+// usage between a time period for a number of hosts (if 0, it will search all hosts),
+// e.g. in pseudo-PromQL:
+//
+// max(
+//	max_over_time(
+// 		{__name__=~'cpu_.*', hostname=~"hostname1|hostname2...|hostnameN"}[12h]
+// 	)
+// ) by (hostname) > 90
+func (d *Devops) HighCPUForHosts(qq query.Query, nHosts int) {
```

Comment:
This seems to be returning very different data than infuxdb -- the 12h grouping here isn't present in the equivalent influx or timescale query. Thus those systems return several hundred MB of JSON whereas Victoria is returning only two data points per host, making it way faster.

Reply:
Thanks for noticing! I removed this case since it is unclear how to support it via MetricsQL/PromQL. The query model assumes step (grouping by time) to be always set. So to satisfy this case we need to pass `log-interval` flag value used for generating data which we don't know when generating queries. I hope other cases are consistent and comparable with other databases. Thank you again!